### PR TITLE
BUG: `stats.levy_stable`: fix ignored parameterization

### LINKED ISSUE #20821: BUG: `stats.levy_stable.rvs`: "S0"-parameterization ignored
### Describe your issue.

When trying to sample from stats.levy_stable.rvs() with levy_stable.parametrization="S0", the returned samples are not transformed, but in the "S1"-parameterization. Line 850 of stats._levy_stable.__init__ does not get executed.
This is likely due to automatically resolving to the frozen distribution in stats._distn_infrastructure:rv_frozen.rvs() [line 493]. 

### Reproducing Code Example

```python
import numpy as np
from scipy.stats import levy_stable

stable_distribution = levy_stable(alpha=1.9, beta=0.1, loc=0.0, scale=1.0)
stable_distribution.parameterization = "S0" #  we could also set this directly as levy_stable.parameterization, makes no difference
samples_s0 = stable_distribution.rvs(size=(1000,1), random_state=49)

stable_distribution.parameterization = "S1"
samples_s1 = stable_distribution.rvs(size=(1000,1), random_state=49)

# due to the needed transformation between "S0" and "S1" parameterizations, the samples should definitely not be identical, but they are
assert not np.all(np.isclose(samples_s0, samples_s1)) # this should hold!
```

### Error message

```shell
No error message
```

### SciPy/NumPy/Python version and system information

<details>

```shell
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "3.0.10",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.15.0",
      "include directory": "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-build-env-o281jlrt\\overlay\\Lib\\site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "windows"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "windows"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "openblas",
      "found": true,
      "version": "0.3.27",
      "detection method": "pkgconfig",
      "include directory": "/c/opt/64/include",
      "lib directory": "/c/opt/64/lib",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS= NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=24",
      "pc file directory": "c:/opt/64/lib/pkgconfig"
    },
    "lapack": {
      "name": "openblas",
      "found": true,
      "version": "0.3.27",
      "detection method": "pkgconfig",
      "include directory": "/c/opt/64/include",
      "lib directory": "/c/opt/64/lib",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS= NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=24",
      "pc file directory": "c:/opt/64/lib/pkgconfig"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.12.0",
      "detection method": "config-tool",
      "include directory": "unknown"
    }
  },
  "Python Information": {
    "path": "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\cibw-run-622t44xc\\cp39-win_amd64\\build\\venv\\Scripts\\python.exe",
    "version": "3.9"
  }
}
```

</details>

### PR DESCRIPTION
**Pull request by Linfye**
#### Reference issue

Solves #20821
#### What does this implement/fix?

I modified the parameters used in the _parameterization function of the levy_stable_gen class. Previously, this function did not read the passed parameters because it used class attributes. By changing it to use instance attributes, this bug can be resolved.
#### Additional information

This is an example:
```
import numpy as np
from scipy.stats import levy_stable

stable_distribution = levy_stable(alpha=1.9, beta=0.1, loc=0.0, scale=1.0)

levy_stable.parameterization = "S0"
samples_s0 = stable_distribution.rvs(size=(1000,1), random_state=49)

levy_stable.parameterization = "S1"
samples_s1 = stable_distribution.rvs(size=(1000,1), random_state=49)

assert not np.all(np.isclose(samples_s0, samples_s1))
```

### CODE REVIEW

**Comment by lucascolley** [comment]
As per https://github.com/scipy/scipy/issues/20821#issuecomment-2273366487, it sounds like this does not solve the issue. I'll close this for now, but feel free to reopen if you can figure it out along the lines of that comment. Thanks for the effort regardless!

**Comment by Linfye** [comment]
@lucascolley I add a test in `test_distributions.py`. Can you review this?

**Comment by lucascolley** [comment]
@ev-br I know it was 11 years ago... but from gh-3245 it seems you have some familiarity with the frozen distribution infrastructure - could you review this?

**Comment by ev-br** [comment]
So the error here is that setting `parametrization` works on a regular instance and fails on a frozen instance, correct?

FWIW, I'm not sure about the fix here. While it does fix an immediate problem, it breaks an invariant that a frozen instance is indeed "frozen" at the creation time and changes made to regular instances afterwards do not propagate via some  "spooky action on a distance". 

The fix here goes precisely the other way: if something somewhere changes the parametrization, the frozen instance is affected. 

The root problem IIUC is that `levy_stable` has a mutable attribute which other distributions do not have. For any other distribution, to change the parametrization you need to create a `rv_continuous` subclass and its instance. This can then be frozen in a regular way. 

So the proposed fix makes a workaround on top of a workaround. Whether this is acceptable from the pragmatic POV, or whether there are better ways to achieve the same effect, I cannot say---it's been a while since I worked on these things :-).

What I have to ask though is whether creating a separate frozen instance does not work: cannot you change the paramerization and then freeze?

**Comment by Linfye** [comment]
In the begining, I tried other ways to pass the parametrization set by the user. But any other modifications to the function `levy_stable_gen` caused thorny problems. So I choose this way. 

Since the parametrization is only used in this function, I guess it would not result in any other problems and the existing tests worked fine. And I noticed the existing tests don't cover all situations as reported by pytest. So I plan to improve this part of the test in the future to find potential problems after this.

**Comment by Linfye** [comment]
@ev-br is there anything else I need to do about this PR?

cc: @lucascolley

**Comment by lucascolley** [comment]
I think we are waiting for someone a bit more familiar with `levy_stable` to give an opinion.

**Comment by Linfye** [comment]
@steppi Could you please take some time to review this code? I noticed you have experience with this before.

This is my first PR to Scipy. Thanks!

**Comment by fancidev** [comment]
I think the fix proposed in this PR "works" but in a somewhat restricted way. The reason why the current code "does not work" is due to the [following line](https://github.com/scipy/scipy/blob/87c46641a8b3b5b47b81de44c07b840468f7ebe7/scipy/stats/_distn_infrastructure.py#L504) in `rv_frozen.__init__()`:

```python
        # create a new instance
        self.dist = dist.__class__(**dist._updated_ctor_param())
```

Note that `dist is levy_stable` and `type(dist) is levy_stable_gen` in the above. So a **copy** of the distribution object (`levy_stable`) is constructed and attached along with the frozen distribution object. That's why code like `scipy.stats.levy_stable.parameterization = "S0"` has no effect on frozen distribution objects. The PR proposed a workaround by essentially always accessing  `levy_stable.parameterization` for *any* distribution object of type `levy_stable_gen`.

An alternative to the proposed workaround is to modify the attributes of the copied distribution object attached to the frozen distribution object. For the test case included with the PR, this would look like the following:

```python
import scipy.stats
import numpy as np

stable_distribution = scipy.stats.levy_stable(alpha=1.9, beta=0.1, loc=0.0, scale=1.0)
stable_distribution.dist.parameterization = "S0"
samples_s0 = stable_distribution.rvs(size=(1000,1), random_state=49)
stable_distribution.dist.parameterization = "S1"
samples_s1 = stable_distribution.rvs(size=(1000,1), random_state=49)
assert not np.all(np.isclose(samples_s0, samples_s1)). # passes without exception
```

**Comment by Linfye** [comment]
This does make more sense, thank you for your analysis and explanation.

**Comment by mdhaber** [comment]
I pushed an alternative fix. Now:
- A frozen distribution `frozen` has a `parameterization` property that is synced with `frozen.dist.parameterization`. Since `frozen.dist` is a distinct instance of `levy_stable_gen` from `stats.levy_stable`, the two `parameterization` attributes are separate. (Even though `parameterization` started as a class variable, it becomes an instance variable when it is set.)
- Also important, I think, but not yet discussed: the initial value of `frozen.dist.parameterization` is the value of `stats.levy_stable.parameterization` whenever the distribution is frozen.

The tests check for these two things.

File: scipy/stats/_levy_stable/__init__.py [hunk 1]
```diff
@@ -816,6 +816,11 @@ class levy_stable_gen(rv_continuous):
     pdf_fft_interpolation_level = 3
     pdf_fft_interpolation_degree = 3
 
+    def freeze(self, *args, **params):
```
Comment (mdhaber) [inline scipy/stats/_levy_stable/__init__.py:819]:
```suggestion
    def __call__(self, *args, **params):
```
This should fix the refguide check failure.

**Comment by Linfye** [comment]
Sorry, I just saw the message. 

As fancidev mentioned, the main problem is that "dist is levy_stable and type(dist) is levy_stable_gen" which cause that the copy of the instance's parameterization won't work. You define a new class `levy_stable_frozen` to accept `parameterization` from the object IIUC. And this solution doesn't change the `levy_stable.parameterization` which seems indeed more practical to me.

**Comment by mdhaber** [comment]
> And this solution doesn't directly change the `levy_stable.parameterization` which seems indeed more practical to me.

If I understand correctly, modifying the `levy_stable.parameterization` was the problem with the approach you had, and it is what this solution avoids. 

Let's forget about the particulars of how this solution working and discuss the goals. Perhaps we are not trying to solve the same problem. Here is what I believe the behavior should be:

1. **Parameterization of un-frozen (`levy_stable`) distribution**:
   - If one sets `levy_stable.parameterization='S0'`, calls to methods of `levy_stable` (e.g. `levy_stable.pdf(1.4, alpha=1.9, beta=0.1, loc=0.0, scale=1.0)`) should use the 'S0' parameterization.
   - If one sets `levy_stable.parameterization='S1'`, calls to methods of `levy_stable` should use the 'S1' parameterization.
2. **Parameterization of frozen distributions**. Suppose one creates a frozen distribution, say `stable_distribution = levy_stable(alpha=1.9, beta=0.1, loc=0.0, scale=1.0)`. Then:
   - If one sets `stable_distribution.parameterization='S0'`, calls to methods of `stable_distribution` (e.g. `stable_distribution.pdf(1.4)`) should use the 'S0' parameterization. 
   - If one sets `stable_distribution.parameterization='S1'`, calls to methods of `stable_distribution` should use the 'S1' parameterization.
3. **Initial parameterization of frozen distributions**. When I instantiate `stable_distribution`, the initial value of `stable_distribution.parameterization` should be the current value of `levy_stable.parameterization`.
4. **Relationship between frozen and un-frozen distribution parameterizations**. After `stable_distribution` is instantiated, changes to the parameterization of `stable_distribution` and `levy_stable` should be independent of one another. If I set `stable_distribution.parameterization`, it should not affect `levy_stable.parameterization`, and vice versa.

Which of these do you agree with, and which do you not agree with (and why)?

**Comment by Linfye** [comment]
The current scipy doc shows [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.levy_stable.html#scipy.stats.levy_stable). So my first PR is in this way.
image

It seems that the original author prefered the first way. But I do agree with the second solution. Because I think it's much more intuitive and convenient which means a beginner usually choose to set the parameter in this way. The user bewit's sample code in #20821 proved this.

And I agree with 3 and 4. These might help to ensure the robustness of users' code.

**Comment by mdhaber** [comment]
@Linfye i can't tell from what you've written here, but are you familiar with the idea of a frozen distribution? Apologies if you are, but just in case:

When I call `stable_distribution = levy_stable(alpha=1.9, beta=0.1, loc=0.0, scale=1.0)`, `stable_distribution` becomes a separate object from `levy_stable`, and I no longer need to pass shape parameters into it's methods; e.g. I can do just `stable_distribution.mean()` and I get the mean for the stable distribution with parameters `alpha=1.9, beta=0.1, loc=0.0, scale=1.0`.

I can have multiple frozen distributions with different shape parameters, e.g. `stable_distribution2 = levy_stable(alpha=2.9, beta=0.9, loc=1.0, scale=2.0)`. Now `stable_distribution2`, `stable_distribution`, and `levy_stable` are all separate objects. The first two remember the (separate) shape parameters they are given, and they can even have their own random state. 

Under the hood, they have in the past used `levy_stable` to do computation because `left_stable` is mostly stateless. But because of the extra `parameterization` attribute (and several others) that are unique to `levy_stable` compared to other distributions, it turns out not to be stateless like all the other distributions in `scipy`. This was overlooked in the design of `levy_stable_gen` and led to the problem reported in gh-20821. The frozen distribution objects used the same parametrization as `levy_stable.parameterization`, and setting the `parameterization` of the frozen distribution had no effect. 

This PR currently fixes the bug and makes everything work as described in points 1-4 above. It sounds like you agree with those, so would you take another look at this? For instance, take a look at the tests - do you agree that they demonstrate ideas 1-4 above? It sounded like you disagreed with something before - if you still do, can you elaborate on what it is?

**Comment by Linfye** [comment]
Actually I wasn’t really familiar with this part before and had some misunderstandings. Thanks for your patience and explaination. Now I can understand what you mean. I didn't understand the connection and difference between frozen distribution and non-frozen distribution before. I search this part on Google and I can understand this feature a little bit. Sorry for this.

I checked the test but didn't run it locally. If there are no errors, I think the solution are reasonable and I agree with the 4 points.

**Comment by mdhaber** [comment]
Great, in that case, think it can go back to @lucascolley, if he is still interested in reviewing this one. If so Lucas - if you agree with the behavior proposed in 1-4 [above](https://github.com/scipy/scipy/pull/21332#issuecomment-2306328941), agree that the tests check that this behavior is satisfied, and think that the code implementing that behavior is clean enough, you're welcome to merge, or I can. Thanks!

**Comment by mdhaber** [comment]
Hey @lucascolley shall I merge this, or would you still like to take a look?

**Review by lucascolley — APPROVED** [review]
Oops sorry, this slipped off my radar!

> if you agree with the behavior proposed in 1-4 https://github.com/scipy/scipy/pull/21332#issuecomment-2306328941, agree that the tests check that this behavior is satisfied, and think that the code implementing that behavior is clean enough, you're welcome to merge, or I can. Thanks!

Yep, all LGTM, thanks @mdhaber and @Linfye for your first contribution to SciPy!

**Pull request merged**